spi/winbond.c

```diff
@@ -285,6 +295,166 @@ static int winbond_is_locked(struct spi_flash *flash)
 	return 0;
 }
 
+static int winbond_sec_read(struct spi_flash *flash, u32 offset, size_t len, void *buf)
+{
+	int ret = 1;
+	u8 cmd[5];
+	u8 reg = (offset >> 8) & 0xFF;
+	u8 addr = offset & 0xFF;
+
+	if (reg != ADDR_W25_SEC1 && reg != ADDR_W25_SEC2 && reg != ADDR_W25_SEC3) {
+		spi_debug("SF: Wrong security register\n");
+		return 1;
+	}
+
+	cmd[0] = CMD_W25_RD_SEC;
+	cmd[1] = 0x0;
+	cmd[2] = reg;
+	cmd[3] = addr;
+	cmd[4] = 0x0; // dummy
+
+	flash->spi->rw = SPI_READ_FLAG;
+	ret = spi_claim_bus(flash->spi);
+	if (ret) {
+		spi_debug("SF: Unable to claim SPI bus\n");
+		return ret;
+	}
+
+	ret = spi_flash_cmd_read(flash->spi, cmd, sizeof(cmd), buf, len);
+	if (ret) {
+		spi_debug("SF: Can't read sec register %d\n", reg >> 4);
+		goto out;
+	}
+
+out:
+	spi_release_bus(flash->spi);
+	return ret;
+}
+
+static int winbond_sec_program(struct spi_flash *flash, u32 offset, size_t len, const void *buf)
+{
+	int ret = 1;
+	u8 cmd[4];
+	u8 reg = (offset >> 8) & 0xFF;
+	u8 addr = offset & 0xFF;
+	u32 tmp_sect_size = flash->sector_size;
+
+	if (reg != ADDR_W25_SEC1 && reg != ADDR_W25_SEC2 && reg != ADDR_W25_SEC3) {
+		spi_debug("SF: Wrong security register\n");
+		return 1;
+	}
+
+	flash->sector_size = 1;
+	ret = spi_flash_cmd_erase(flash, CMD_W25_ER_SEC, offset & (0xFF << 8), 1);
+	flash->sector_size = tmp_sect_size;
+	if (ret) {
+		spi_debug("SF: Can't erase sec reg\n");
+		return ret;
+	}
+
+	flash->spi->rw = SPI_WRITE_FLAG;
+	ret = spi_claim_bus(flash->spi);
+	if (ret) {
+		spi_debug("SF: Unable to claim SPI bus\n");
+		return ret;
+	}
+
+	ret = spi_flash_cmd(flash->spi, CMD_W25_WREN, NULL, 0);
+	if (ret) {
+		spi_debug("SF: Enabling Write failed\n");
+		goto out;
+	}
+
+	cmd[0] = CMD_W25_WR_SEC;
+	cmd[1] = 0x0;
+	cmd[2] = reg;
+	cmd[3] = addr;
+	ret = spi_flash_cmd_write(flash->spi, cmd, sizeof(cmd), buf, len);
+	if (ret) {
+		spi_debug("SF: Can't write to sec register %d\n", reg >> 4);
+		goto out;
+	}
+
+	ret = spi_flash_cmd_wait_ready(flash, SPI_FLASH_PROG_TIMEOUT);
+	if (ret) {
+		spi_debug("SF: Programming sec register failed - timeout\n");
+		goto out;
+	}
+
+out:
+	spi_release_bus(flash->spi);
+	return ret;
+}
+
+static int winbond_sec_sts(struct spi_flash *flash)
+{
+	u8 status = 0;
+	int ret;
+
+	ret = spi_flash_cmd(flash->spi, CMD_W25_RDSR2, &status, 1);
+	if (ret) {
+		return -1;
+	}
+
+	status = status & (REG_W25_LB1 | REG_W25_LB2 | REG_W25_LB3) >> 3;
+
+	return status;
+}
+
+static int winbond_sec_lock(struct spi_flash *flash, u8 reg)
+{
+	int ret;
+	u8 status = 0;
+	u8 cmd;
+
+	flash->spi->rw = SPI_WRITE_FLAG;
+	ret = spi_claim_bus(flash->spi);
+	if (ret) {
+		spi_debug("SF: Unable to claim SPI bus\n");
+		return ret;
+	}
+
+	cmd = CMD_W25_RDSR2;
+	ret = spi_flash_cmd(flash->spi, cmd, &status, sizeof(status));
```

Comment:
@mek-x `spi_flash_cmd` calling patterns are not consistent. Sometimes we use `cmd` sometimes we provide it directly. We shall be more consistent.

Reply:
it's consistent with other functions. Problem is, that spi_flash_cmd expects value and spi_flash_cmd_write expects pointer